generator/integration_tests/golden/internal/golden_kitchen_sink_stub_factory.gcpcxx.pb.cc

```diff
@@ -28,11 +34,34 @@ namespace cloud {
 namespace golden_internal {
 inline namespace GOOGLE_CLOUD_CPP_GENERATED_NS {
 
+internal::Options ResolveGoldenKitchenSinkOptions(internal::Options options) {
+  if (!options.has<internal::EndpointOption>()) {
+    auto env = internal::GetEnv("GOOGLE_CLOUD_CPP_GOLDEN_KITCHEN_SINK_ENDPOINT");
+    options.set<internal::EndpointOption>(env ? *env : "goldenkitchensink.googleapis.com");
+  }
+  if (!options.has<internal::GrpcCredentialOption>()) {
+    options.set<internal::GrpcCredentialOption>(grpc::GoogleDefaultCredentials());
+  }
+  if (!options.has<internal::GrpcBackgroundThreadsFactoryOption>()) {
+    options.set<internal::GrpcBackgroundThreadsFactoryOption>(
+        internal::DefaultBackgroundThreadsFactory);
+  }
+  if (!options.has<internal::GrpcNumChannelsOption>()) {
+    options.set<internal::GrpcNumChannelsOption>(4);
+  }
+  auto& products = options.lookup<internal::UserAgentProductsOption>();
+  products.insert(products.begin(), google::cloud::internal::UserAgentPrefix());
+
+  return options;
+}
+
 std::shared_ptr<GoldenKitchenSinkStub>
-CreateDefaultGoldenKitchenSinkStub(golden::GoldenKitchenSinkConnectionOptions const& options) {
-  auto channel =
-      grpc::CreateCustomChannel(options.endpoint(), options.credentials(),
-                                options.CreateChannelArguments());
+CreateDefaultGoldenKitchenSinkStub(internal::Options options) {
+  options = ResolveGoldenKitchenSinkOptions(std::move(options));
```

Comment:
There should be no need to call `ResolveGoldenKitchenSinkOptions()` here, because the user-facing "MakeConnection" factory would have already set all the defaults as appropriate. Therefore, this function can/should accept `internal::Options const& opts` by const-ref, because it's only reading from the object.

FWIW: The approach I've taken with Spanner is this:

- Spanner's user-facing "make connection" factory functions accept `internal::Options opts` *by value*, because they need a copy because they're going to tweak the options to set any missing defaults appropriately. This is the only place -- in the one "connection factory layer" -- where defaults are resolved.
- All other *non-user-facing* functions that accept an `Optoins` do so by const-ref, because they don't need to modify anything.